### ISSUE TITLE
Library is too heavy

### ISSUE THREAD

**Issue by kalemas** [krathjen/studiolibrary#198]
Hi, our Library has become hard to operate with one day. I did not made a deep research but profiling tell following:

```
14:46:18 inhouse.apps.studiolibrary_runner: run ()
         244290 function calls (243367 primitive calls) in 20.694 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   20.694   20.694 d:\dev\bundle\python\inhouse\apps\studiolibrary_runner.py:24(run)
        1    0.005    0.005   20.687   20.687 d:\dev\bundle\python\studiolibrary\main.py:17(main)
        1    0.001    0.001   20.655   20.655 d:\dev\bundle\python\studiolibrary\packages\studiolibrarymaya\main.py:17(main)
        1    0.000    0.000   20.634   20.634 d:\dev\bundle\python\studiolibrary\librarywidget.py:118(instance)
        1    0.000    0.000   20.617   20.617 d:\dev\bundle\python\studiolibrary\packages\studiolibrarymaya\mayalibrarywidget.py:96(show)
        1    0.002    0.002   20.610   20.610 C:\Program Files\Autodesk\Maya2018\Python\lib\site-packages\maya\app\general\mayaMixin.py:415(show)
    52/34    0.000    0.000   20.598    0.606 d:\dev\bundle\python\studiolibrary\librarywidget.py:1939(event)
        1    0.000    0.000   20.594   20.594 d:\dev\bundle\python\studiolibrary\librarywidget.py:1980(showEvent)
      3/1    0.000    0.000   20.594   20.594 d:\dev\bundle\python\studiolibrary\packages\studioqt\decorators.py:21(wrapped)
        1    0.000    0.000   20.593   20.593 d:\dev\bundle\python\studiolibrary\librarywidget.py:1852(loadSettings)
        1    0.000    0.000   20.583   20.583 d:\dev\bundle\python\studiolibrary\librarywidget.py:1750(setSettings)
        3    0.000    0.000   20.535    6.845 d:\dev\bundle\python\studiolibrary\librarywidget.py:556(refresh)
        1    0.000    0.000   20.518   20.518 d:\dev\bundle\python\studiolibrary\librarywidget.py:614(refreshFolders)
        1    0.014    0.014   20.518   20.518 d:\dev\bundle\python\studiolibrary\librarywidget.py:630(updateFolders)
     1535    0.007    0.000   20.349    0.013 d:\dev\bundle\python\studiolibrary\cmds.py:224(findItems)
     1548    0.011    0.000   13.204    0.009 d:\dev\bundle\python\studiolibrary\cmds.py:153(itemFromPath)
     1534    0.006    0.000   12.297    0.008 d:\dev\bundle\python\studiolibrary\libraryitem.py:127(__init__)
     1534    0.004    0.000   12.270    0.008 d:\dev\bundle\python\studiolibrary\libraryitem.py:404(setPath)
     1534    0.022    0.000   12.260    0.008 d:\dev\bundle\python\studiolibrary\libraryitem.py:422(updateData)
     3310   12.153    0.004   12.153    0.004 {nt.stat}
     1308    0.006    0.000   10.591    0.008 d:\dev\bundle\python\studiolibrary\packages\studiolibrarymaya\baseitem.py:71(__init__)
  789/172    0.009    0.000    7.118    0.041 C:\Program Files\Autodesk\Maya2018\bin\python27.zip\os.py:209(walk)
     1776    7.042    0.004    7.042    0.004 {nt._isdir}
     1535    0.004    0.000    6.081    0.004 C:\Program Files\Autodesk\Maya2018\bin\python27.zip\genericpath.py:60(getmtime)
     1536    0.004    0.000    6.075    0.004 C:\Program Files\Autodesk\Maya2018\bin\python27.zip\genericpath.py:23(exists)
      721    0.005    0.000    5.712    0.008 d:\dev\bundle\python\studiolibrary\packages\studiolibrarymaya\poseitem.py:80(__init__)
      282    0.001    0.000    2.538    0.009 d:\dev\bundle\python\studiolibrary\packages\studiolibrarymaya\setsitem.py:33(__init__)
      282    0.001    0.000    2.214    0.008 d:\dev\bundle\python\studiolibrary\packages\studiolibrarymaya\animitem.py:91(__init__)
      226    0.001    0.000    1.705    0.008 d:\dev\bundle\python\studiolibrary\folderitem.py:65(updateData)
```

This launch displayed root library folders in side panel and empty item view. Recursive search is disabled. So can consider that there is an issue with updating Library. 20 seconds in `findItems` and 1300 calls of `baseitem.__init__` to display empty view... And i have same when i create an item. Resetting settings does not fix anything.

**Comment by krathjen** [comment]
Hey @kalemas,

Yea this seems very slow. I test with over 1000 items which for me loads within 1sec. This could be a network issue when Studio Library walks the directory. The following branch supports scandir which might help. You will need to install scandir though.

https://github.com/benhoyt/scandir

Can you test the following branch?

https://github.com/krathjen/studiolibrary/tree/imp/librarymodel

Let me know if this helps.

Thanks,

Kurt

**Comment by kalemas** [comment]
Better, but only in sense of `os.walk` (we strip profiler results with less than 0.1 cumtime here) )):
```
19:25:51 inhouse.apps.studiolibrary_runner: run ()
         309432 function calls (308483 primitive calls) in 24.276 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   24.276   24.276 d:\dev\bundle\python\inhouse\apps\studiolibrary_runner.py:24(run)
        1    0.005    0.005   24.269   24.269 d:\dev\bundle\python\studiolibrary\main.py:17(main)
        1    0.001    0.001   24.237   24.237 d:\dev\bundle\python\studiolibrary\packages\studiolibrarymaya\main.py:17(main)
        1    0.000    0.000   24.216   24.216 d:\dev\bundle\python\studiolibrary\librarywidget.py:119(instance)
        1    0.000    0.000   24.202   24.202 d:\dev\bundle\python\studiolibrary\packages\studiolibrarymaya\mayalibrarywidget.py:96(show)
        1    0.006    0.006   24.191   24.191 C:\Program Files\Autodesk\Maya2018\Python\lib\site-packages\maya\app\general\mayaMixin.py:415(show)
    63/36    0.000    0.000   24.169    0.671 d:\dev\bundle\python\studiolibrary\librarywidget.py:1864(event)
        1    0.000    0.000   24.165   24.165 d:\dev\bundle\python\studiolibrary\librarywidget.py:1905(showEvent)
      5/1    0.000    0.000   24.165   24.165 d:\dev\bundle\python\studiolibrary\packages\studioqt\decorators.py:21(wrapped)
        1    0.000    0.000   24.164   24.164 d:\dev\bundle\python\studiolibrary\librarywidget.py:1777(loadSettings)
        1    0.000    0.000   24.154   24.154 d:\dev\bundle\python\studiolibrary\librarywidget.py:1672(setSettings)
        3    0.000    0.000   24.096    8.032 d:\dev\bundle\python\studiolibrary\librarywidget.py:611(refresh)
        1    0.000    0.000   24.096   24.096 d:\dev\bundle\python\studiolibrary\librarywidget.py:620(update)
        1    0.000    0.000   24.092   24.092 d:\dev\bundle\python\studiolibrary\librarywidget.py:673(refreshFolders)
        1    0.001    0.001   24.091   24.091 d:\dev\bundle\python\studiolibrary\librarywidget.py:689(updateFolders)
        2    0.001    0.000   23.879   11.939 d:\dev\bundle\python\studiolibrary\librarymodel.py:284(createItems)
     2600    0.037    0.000   21.943    0.008 d:\dev\bundle\python\studiolibrary\libraryitem.py:425(updateData)
     5464   21.759    0.004   21.759    0.004 {nt.stat}
     1545    0.008    0.000   14.501    0.009 d:\dev\bundle\python\studiolibrary\cmds.py:234(findItems)
     1568    0.011    0.000   13.544    0.009 d:\dev\bundle\python\studiolibrary\cmds.py:163(itemFromPath)
     1544    0.007    0.000   12.647    0.008 d:\dev\bundle\python\studiolibrary\libraryitem.py:127(__init__)
     1544    0.004    0.000   12.609    0.008 d:\dev\bundle\python\studiolibrary\libraryitem.py:407(setPath)
     2605    0.006    0.000   11.152    0.004 C:\Program Files\Autodesk\Maya2018\bin\python27.zip\genericpath.py:23(exists)
     1318    0.006    0.000   10.964    0.008 d:\dev\bundle\python\studiolibrary\packages\studiolibrarymaya\baseitem.py:71(__init__)
     2604    0.007    0.000   10.615    0.004 C:\Program Files\Autodesk\Maya2018\bin\python27.zip\genericpath.py:60(getmtime)
        1    0.005    0.005    9.369    9.369 d:\dev\bundle\python\studiolibrary\librarymodel.py:320(loadItemData)
      730    0.004    0.000    6.062    0.008 d:\dev\bundle\python\studiolibrary\packages\studiolibrarymaya\poseitem.py:77(__init__)
      343    0.001    0.000    2.736    0.008 d:\dev\bundle\python\studiolibrary\folderitem.py:65(updateData)
      282    0.001    0.000    2.496    0.009 d:\dev\bundle\python\studiolibrary\packages\studiolibrarymaya\setsitem.py:32(__init__)
      283    0.001    0.000    2.275    0.008 d:\dev\bundle\python\studiolibrary\packages\studiolibrarymaya\animitem.py:91(__init__)
```

**Comment by krathjen** [comment]
What kind of speed improvements do you see when you comment out the following code from L442 to L447?

https://github.com/krathjen/studiolibrary/blob/a778e802e512c1d9d4e52cf094b3a656917ebd66/libraryitem.py#L442

```
        # if os.path.exists(path):
            # modified = os.path.getmtime(path)
            # timeAgo = studiolibrary.timeAgo(modified)

            # self.setText("Modified", timeAgo)
            # self.setSortText("Modified", str(modified))
```

**Comment by kalemas** [comment]
First call was 6.5 seconds, look like scandir caches its data.

```
21:46:54 inhouse.apps.studiolibrary_runner: run ()
         332823 function calls (331851 primitive calls) in 2.564 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    2.564    2.564 d:\dev\bundle\python\inhouse\apps\studiolibrary_runner.py:24(run)
        1    0.004    0.004    2.557    2.557 d:\dev\bundle\python\studiolibrary\main.py:17(main)
        1    0.001    0.001    2.526    2.526 d:\dev\bundle\python\studiolibrary\packages\studiolibrarymaya\main.py:17(main)
        1    0.000    0.000    2.507    2.507 d:\dev\bundle\python\studiolibrary\librarywidget.py:119(instance)
        1    0.000    0.000    2.493    2.493 d:\dev\bundle\python\studiolibrary\packages\studiolibrarymaya\mayalibrarywidget.py:96(show)
        1    0.006    0.006    2.464    2.464 C:\Program Files\Autodesk\Maya2018\Python\lib\site-packages\maya\app\general\mayaMixin.py:415(show)
    63/36    0.000    0.000    2.447    0.068 d:\dev\bundle\python\studiolibrary\librarywidget.py:1864(event)
        1    0.000    0.000    2.443    2.443 d:\dev\bundle\python\studiolibrary\librarywidget.py:1905(showEvent)
      9/1    0.000    0.000    2.443    2.443 d:\dev\bundle\python\studiolibrary\packages\studioqt\decorators.py:21(wrapped)
        1    0.000    0.000    2.442    2.442 d:\dev\bundle\python\studiolibrary\librarywidget.py:1777(loadSettings)
        1    0.000    0.000    2.431    2.431 d:\dev\bundle\python\studiolibrary\librarywidget.py:1672(setSettings)
        3    0.000    0.000    2.069    0.690 d:\dev\bundle\python\studiolibrary\librarywidget.py:611(refresh)
        1    0.000    0.000    2.069    2.069 d:\dev\bundle\python\studiolibrary\librarywidget.py:620(update)
        1    0.000    0.000    2.066    2.066 d:\dev\bundle\python\studiolibrary\librarywidget.py:673(refreshFolders)
        1    0.000    0.000    2.066    2.066 d:\dev\bundle\python\studiolibrary\librarywidget.py:689(updateFolders)
        3    0.000    0.000    1.854    0.618 d:\dev\bundle\python\studiolibrary\librarymodel.py:284(createItems)
     1555    0.004    0.000    1.795    0.001 d:\dev\bundle\python\studiolibrary\cmds.py:234(findItems)
```

**Comment by krathjen** [comment]
Hey @kalemas,

Which branch/version were those results from?

I pushed some more changes to the following branch a couple of days ago which might also help.

https://github.com/krathjen/studiolibrary/tree/imp/librarymodel

Thanks,

Kurt

**Comment by kalemas** [comment]
@krathjen its a778e802e512c1d9d4e52cf094b3a656917ebd66 with commented lines you pointed.

**Comment by krathjen** [comment]
Hey @kalemas,

You may also find that the following commit has some improvements aswell.

https://github.com/krathjen/studiolibrary/commit/e603c4bd22900775946724ea9c1f7c278c3b1284

What is the speed difference with and without the commented lines L442 to L447?

Thanks,

Kurt

**Comment by kalemas** [comment]
@krathjen 
20 on `master`
24 on `imp/librarymodel`
2.5 with commented lines L442 to L447

so, it is ten times faster with lines commented

**Comment by krathjen** [comment]
Thanks, @kalemas, this is good to know.

**Comment by kalemas** [comment]
now it loads quite fast but we do not see items in f8238d58bc24afca2c6ad00dfdfdf9ed43577ef7 without recursive search enabled
![2018-08-23_18-12-32](https://user-images.githubusercontent.com/638393/44534427-6d46df80-a700-11e8-946d-0539940fe4d7.png)

**Comment by krathjen** [comment]
Hey @kalemas,

Closing this ticket as it should now be fixed in version 2.4.0.b15.

Feel free to reopen the ticket or create a new one if you're still having issues.

Thanks,

Kurt
